### ISSUE TITLE
'opencv2/core/version.hpp' file not found apple M1

### ISSUE THREAD

**Issue by pablovela5620** [google/mediapipe#3440]
Following up on issues #2805 and #2519 I am also getting the error 'opencv2/core/version.hpp' on my M1 Mac. when following
`Building C++ command-line example apps`

I made sure to install Opencv via brew, as well as doing a soft like similar to #2519
`sudo ln -sf /opt/homebrew/Cellar/opencv@3/3.4.14_3 /usr/local/opt/opencv@3`

I ran the example build command
`bazel build -c opt --define MEDIAPIPE_DISABLE_GPU=1 mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu`

I get the following error
```
ERROR: /Users/pablovela/0Dev/repos/mediapipe/mediapipe/framework/formats/BUILD:205:11: Compiling mediapipe/framework/formats/location.cc failed: (Exit 1): sandbox-exec failed: error executing command 
  (cd /private/var/tmp/_bazel_pablovela/ead135855659b22f134b7cba96953e8c/sandbox/darwin-sandbox/6865/execroot/mediapipe && \
  exec env - \
    PATH=/Users/pablovela/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.0.0-darwin-arm64/bin:/Users/pablovela/miniforge3/envs/mp/bin:/opt/homebrew/opt/opencv@3/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/pablovela/miniforge3/condabin:/opt/homebrew/bin:/opt/homebrew/sbin \
    PWD=/proc/self/cwd \
    TMPDIR=/var/folders/tf/rbxr_mz11n19ph5s2tzkqr580000gn/T/ \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_pablovela/ead135855659b22f134b7cba96953e8c/sandbox/darwin-sandbox/6865/sandbox.sb /var/tmp/_bazel_pablovela/install/c8706c49991751b43417f135b97a0616/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/darwin_arm64-opt/bin/mediapipe/framework/formats/_objs/location/location.d '-frandom-seed=bazel-out/darwin_arm64-opt/bin/mediapipe/framework/formats/_objs/location/location.o' '-DMEDIAPIPE_DISABLE_GPU=1' -iquote . -iquote bazel-out/darwin_arm64-opt/bin -iquote external/com_google_protobuf -iquote bazel-out/darwin_arm64-opt/bin/external/com_google_protobuf -iquote external/zlib -iquote bazel-out/darwin_arm64-opt/bin/external/zlib -iquote external/com_google_absl -iquote bazel-out/darwin_arm64-opt/bin/external/com_google_absl -iquote external/com_github_glog_glog -iquote bazel-out/darwin_arm64-opt/bin/external/com_github_glog_glog -iquote external/com_github_gflags_gflags -iquote bazel-out/darwin_arm64-opt/bin/external/com_github_gflags_gflags -iquote external/macos_opencv -iquote bazel-out/darwin_arm64-opt/bin/external/macos_opencv -Ibazel-out/darwin_arm64-opt/bin/external/com_github_glog_glog/_virtual_includes/default_glog_headers -Ibazel-out/darwin_arm64-opt/bin/external/com_github_gflags_gflags/_virtual_includes/gflags -isystem external/com_google_protobuf/src -isystem bazel-out/darwin_arm64-opt/bin/external/com_google_protobuf/src -isystem external/zlib -isystem bazel-out/darwin_arm64-opt/bin/external/zlib -isystem external/macos_opencv/opt/opencv@3/include -isystem bazel-out/darwin_arm64-opt/bin/external/macos_opencv/opt/opencv@3/include -w '-std=c++17' -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c mediapipe/framework/formats/location.cc -o bazel-out/darwin_arm64-opt/bin/mediapipe/framework/formats/_objs/location/location.o)
In file included from mediapipe/framework/formats/location.cc:15:
In file included from ./mediapipe/framework/formats/location.h:44:
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
#include <opencv2/core/version.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 51.530s, Critical Path: 36.51s
INFO: 1834 processes: 956 internal, 878 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

I'm not really sure whats wrong here, I also tried replacing `opencv_linux.BUILD` with `opencv_macos.BUILD` but got the same error

any help would be appreciated, I have no problem building on my intel based mac, this is specific to the M1

**Comment by sureshdagooglecom** [comment]
Hi @pablovela5620 ,
Could you provide steps to reproduce this issue.

**Comment by tso996** [comment]
You may already know this but it could be because your m1 mac is not being recognised as an Apple OS by bazel and is choosing the default configuration in the BUILD file and as a result, some of the necessary components for building in Apple os aren't being built.
A step in the right direction would be to open the BUILD file in the folder of each file that is throwing the error and adding the Apple specific dependencies to the default dependency. Of course, this could result in undefined behaviours when you build for other platforms without reverting to the original settings but it will get it to work properly for the current situation.

For example,
```In file included from ./mediapipe/framework/formats/location.h:44:
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
#include <opencv2/core/version.hpp>
```
To fix this problem, you may need to go into /mediapipe/framework/formats/BUILD and put any opencv related dependency for mediapipe:apple into the default dependency value. 
```
deps = select({
        "//conditions:default": [
            "//mediapipe/framework/deps:image_resizer",
            "//mediapipe/framework/port:opencv_core",
        ],
    }),
```
Are you sure that you are using the latest MediaPipe release? I was under the impression that the MediaPipe team had already sorted out the compatibility issues with the new M1 macs recently.

**Comment by FereshtehFeiz** [comment]
I have the same error: 
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
I want to run:
bazel run --define MEDIAPIPE_DISABLE_GPU=1 mediapipe/examples/desktop/iris_tracking:iris_depth_from_image_desktop
on Mac M1
I don't know how to fix it? while I can run the HelloWorld.

**Comment by aadityasp** [comment]
>  while I can run the HelloWorld.
Samee! were you able to fix the error?

**Comment by LluisColom** [comment]
> > You may already know this but it could be because your m1 mac is not being recognised as an Apple OS by bazel and is choosing the default configuration in the BUILD file and as a result, some of the necessary components for building in Apple os aren't being built. A step in the right direction would be to open the BUILD file in the folder of each file that is throwing the error and adding the Apple specific dependencies to the default dependency. Of course, this could result in undefined behaviour when you build for other platforms without reverting to the original settings but it will get it to work properly for the current situation.
> > For example,
> > ```
> > ./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
> > #include <opencv2/core/version.hpp>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To fix this problem, you may need to go into /mediapipe/framework/formats/BUILD and put any opencv related dependency for mediapipe:apple into the default dependency value.
> > ```
> > deps = select({
> >         "//conditions:default": [
> >             "//mediapipe/framework/deps:image_resizer",
> >             "//mediapipe/framework/port:opencv_core",
> >         ],
> >     }),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Are you sure that you are using the latest MediaPipe release? I was under the impression that the MediaPipe team had already sorted out the compatibility issues with the new M1 macs recently.
> 
> So it looks like the issue is at least in part to what you specify here of deps not being correctly pointed to, I used the patches found in the [mediapipe-silicon repo](https://github.com/cansik/mediapipe-silicon) to fix this as well as some problems with `xcode-locator` still being x64 (more info [here](https://github.com/cansik/mediapipe-silicon/issues/2#issuecomment-1160685659)). Lastly I think at least in part it may had to do with not having my OS at 12.4 (though im not really sure about this). Hopefully this helps anyone else that meets this problem.

Hello,

I would not personally recommend to use 3rd parties software to run MediaPipe, as in future releases the changes applied in this repo could be broken (no longer useful).

For this reason, if running into this error using a M1 powered device:
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
#include <opencv2/core/version.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~

I suggest you to apply the following changes to the MediaPipe code:

diff --git a/WORKSPACE b/WORKSPACE
index 146916c..28359c0 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -266,13 +266,13 @@ new_local_repository(
     # For local MacOS builds, the path should point to an opencv@3 installation.
     # If you edit the path here, you will also need to update the corresponding
     # prefix in "opencv_macos.BUILD".
-    path = "/usr/local",
+    path = "/opt/homebrew/",
 )
 
 new_local_repository(
     name = "macos_ffmpeg",
     build_file = "@//third_party:ffmpeg_macos.BUILD",
-    path = "/usr/local/opt/ffmpeg",
+    path = "/opt/homebrew/opt/ffmpeg",
 )

Basically, the problem here is related with the fact that Homebrew is installing OpenCV and all its files in a folder different than the one specified in the WORKSPACE file, so when building Bazel cannot find OpenCV.

As of October 3rd 2022, I'm using the latest version of Mediapipe in my Macbook Pro 14 (M1 pro) and the changes explained here allow MediaPipe models to run successfully.

**Comment by anz-ryu** [comment]
I have the same issue of this.

I am using the master(12f72f0) of the mediapipe repository. (Am I wrong?)

```
ERROR: /Users/user/Documents/IdeaProject/mediapipe/mediapipe/calculators/tensor/BUILD:916:11: Compiling mediapipe/calculators/tensor/image_to_tensor_converter_opencv.cc failed: (Aborted): wrapped_clang_pp failed: error executing command external/local_config_cc/wrapped_clang_pp '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -g0 -O2 -DNDEBUG ... (remaining 75 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from mediapipe/calculators/tensor/image_to_tensor_converter_opencv.cc:25:
In file included from ./mediapipe/framework/formats/image_opencv.h:20:
./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
#include <opencv2/core/version.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Error in child process '/usr/bin/xcrun'. 1
Target //mediapipe/modules/face_detection:face_detection_short_range_cpu failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 31.458s, Critical Path: 27.17s
INFO: 296 processes: 15 internal, 281 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

**Comment by suyashcjoshi** [comment]
> > > You may already know this but it could be because your m1 mac is not being recognised as an Apple OS by bazel and is choosing the default configuration in the BUILD file and as a result, some of the necessary components for building in Apple os aren't being built. A step in the right direction would be to open the BUILD file in the folder of each file that is throwing the error and adding the Apple specific dependencies to the default dependency. Of course, this could result in undefined behaviour when you build for other platforms without reverting to the original settings but it will get it to work properly for the current situation.
> > > For example,
> > > ```
> > > ./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found
> > > #include <opencv2/core/version.hpp>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > To fix this problem, you may need to go into /mediapipe/framework/formats/BUILD and put any opencv related dependency for mediapipe:apple into the default dependency value.
> > > ```
> > > deps = select({
> > >         "//conditions:default": [
> > >             "//mediapipe/framework/deps:image_resizer",
> > >             "//mediapipe/framework/port:opencv_core",
> > >         ],
> > >     }),
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Are you sure that you are using the latest MediaPipe release? I was under the impression that the MediaPipe team had already sorted out the compatibility issues with the new M1 macs recently.
> > 
> > 
> > So it looks like the issue is at least in part to what you specify here of deps not being correctly pointed to, I used the patches found in the [mediapipe-silicon repo](https://github.com/cansik/mediapipe-silicon) to fix this as well as some problems with `xcode-locator` still being x64 (more info [here](https://github.com/cansik/mediapipe-silicon/issues/2#issuecomment-1160685659)). Lastly I think at least in part it may had to do with not having my OS at 12.4 (though im not really sure about this). Hopefully this helps anyone else that meets this problem.
> 
> Hello,
> 
> I would not personally recommend to use 3rd parties software to run MediaPipe, as in future releases the changes applied in this repo could be broken (no longer useful).
> 
> For this reason, if running into this error using a M1 powered device: ./mediapipe/framework/port/opencv_core_inc.h:18:10: fatal error: 'opencv2/core/version.hpp' file not found #include <opencv2/core/version.hpp> ^~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> I suggest you to apply the following changes to the MediaPipe code:
> 
> diff --git a/WORKSPACE b/WORKSPACE index 146916c..28359c0 100644 --- a/WORKSPACE +++ b/WORKSPACE @@ -266,13 +266,13 @@ new_local_repository( # For local MacOS builds, the path should point to an opencv@3 installation. # If you edit the path here, you will also need to update the corresponding # prefix in "opencv_macos.BUILD". -- path = "/usr/local", ++ path = "/opt/homebrew/", )
> 
> new_local_repository( name = "macos_ffmpeg", build_file = "@//third_party:ffmpeg_macos.BUILD", -- path = "/usr/local/opt/ffmpeg", ++ path = "/opt/homebrew/opt/ffmpeg", )
> 
> Basically, the problem here is related with the fact that Homebrew is installing OpenCV and all its files in a folder different than the one specified in the WORKSPACE file, so when building Bazel cannot find OpenCV.
> 
> As of October 3rd 2022, I'm using the latest version of Mediapipe in my Macbook Pro 14 (M1 pro) and the changes explained here allow MediaPipe models to run successfully.

I had the same problem, this fixed it. I want to submit a PR but don't know if it will break for Non Apple Silicon machines as I don't have a wat to test on that. Seems like the change is mostly updating homebrew path which got updated as mentioned here: https://brew.sh/2021/02/05/homebrew-3.0.0/
